cmds/switch_root/switch_root.go

```diff
@@ -23,140 +24,201 @@ func usage() string {
+			if err != nil {
+				// just delete files
+				if err := unlinkat(fd, name, 0); err != nil {
+					return err
+				}
+			} else {
+				defer syscall.Close(namefd)
+
+				var nameStatT syscall.Stat_t
+				if err := fstatat(fd, name, &nameStatT, unix.AT_SYMLINK_NOFOLLOW); err != nil {
+					return err
+				}
+
+				if fdDev, err := getDev(fd); err != nil {
+					return err
+				} else if fdDev != nameStatT.Dev {
+					// on different filesystem, don't go there
+					continue
+				}
+
+				// not actually a dir, but a symlink to a dir, treat like file and remove
+				if isSymlink(nameStatT.Mode) {
+					if err := unlinkat(fd, name, 0); err != nil {
+						return err
+					}
+					continue
+				}
+
+				// actual recursion
+				if err := recursiveDelete(namefd); err != nil {
+					return err
+				}
+				// back from recursion, dir is not empty, delete
+				if err := unlinkat(fd, name, AT_REMOVEDIR); err != nil {
+					return err
+				}
+			}
 		}
 	}
-
 	return nil
 }
 
-// execCommand will run the executable at "path" with PID 1.
+// do a proper exec to retain PID 1.
 // It returns an error if the command exits incorrectly.
 func execCommand(path string) error {
-	cmd := exec.Command(path)
-	cmd.Stdin = os.Stdin
-	cmd.Stderr = os.Stderr
-	cmd.Stdout = os.Stdout
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Setctty:    true,
-		Setsid:     true,
-		Cloneflags: syscall.CLONE_THREAD | syscall.CLONE_NEWPID,
-	}
-
-	return cmd.Run()
+	return syscall.Exec(path, []string{path}, []string{})
 }
 
-// specialFS creates and mounts proc, sys and dev at the root level.
-func specialFS() error {
-
-	if err := syscall.Mkdir("/proc", 0); err != nil {
-		return err
+// check if dir is empty
+func isEmpty(name string) (bool, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return false, err
 	}
+	defer f.Close()
 
-	if err := syscall.Mkdir("/sys", 0); err != nil {
-		return err
+	_, err = f.Readdirnames(1)
+	if err == io.EOF { // this means emtpy
+		return true, nil
 	}
+	return false, err
+}
 
-	if err := syscall.Mkdir("/dev", 0); err != nil {
-		return err
+// check if file exists
+func isExists(name string) (bool, error) {
+	if _, err := os.Stat(name); err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
 	}
+	return true, nil
+}
 
-	if err := syscall.Mount("", "/proc", "proc", 0, ""); err != nil {
-		return err
-	}
+// moves a mountpoint
+func moveMount(oldPath string, newPath string) error {
+	return syscall.Mount(oldPath, newPath, "", syscall.MS_MOVE, "")
+}
 
-	if err := syscall.Mount("", "/sys", "sysfs", 0, ""); err != nil {
-		return err
-	}
+// specialFS creates and mounts proc, sys, and dev.
+func specialFS(newRoot string) error {
+	var mounts = []string{"/dev", "/proc", "/sys", "/run"}
 
-	return syscall.Mount("", "/dev", "devtmpfs", 0, "")
+	for _, mount := range mounts {
+		path := newRoot + mount
+		// skip all mounting if the dir does not exists
+		if exists, err := isExists(mount); err != nil {
```

Comment:
I feel like `isExists` doesn't actually add any value:

```go
if _, err := os.Stat(mount); os.IsNotExist(err) {
  fmt.Println("switch_root: skipping", mount)
  continue
} else if err != nil {
  return err
}
```

Same thing in the same number of lines, but without the isExists function.